### ISSUE TITLE
static call to invalid vtable

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#20676]
Discovered this ICE in the rollup that I'm preparing:

```rust
fn main() {
    let a: &fmt::Show = &1i;     
    format!("{:?}", a);
}
```

```
error: internal compiler error: static call to invalid vtable: VtableObject(VtableObject(object_ty=core::fmt::Show + 'static))
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: http://doc.rust-lang.org/complement-bugreport.html
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /home/alex/code/rust2/src/libsyntax/diagnostic.rs:182
```

**Comment by mzabaluev** [comment]
In Grust, I tried to change [this line](https://github.com/mzabaluev/grust/blob/5419fe1f80500b2ca6c7567c56aab2f81b326b41/src/callback.rs#L48) to:
```rust
        AsyncClosure::invoke(self.closure, args)
```
The crate compiles without an issue, but compiling another crate where this is used results in an ICE that looks to be this one.

**Comment by nikomatsakis** [comment]
Looks like I missed a case in https://github.com/rust-lang/rust/pull/20341

**Comment by nikomatsakis** [comment]
Nominating for 1.0 beta.

**Comment by shepmaster** [comment]
Are there any known workarounds for this problem? I ran into this in a hard way trying to update to the newest nightlies.

**Comment by nikomatsakis** [comment]
@shepmaster not sure, I'll try to get a PR up today

**Comment by nikomatsakis** [comment]
@shepmaster a new type wrapper around the `Show` object would probably work, ping me on irc (`nmatsakis`) for more details

**Comment by outersky** [comment]
I just wanna run iron example on MAC 10.10 with rust-nightly, the same error occurs, I don't know if it's of the same case:

error: internal compiler error: static call to invalid vtable: VtableObject(VtableObject(object_ty=error::Error + 'static))

**Comment by barosl** [comment]
This is also blocking Iron, which itself compiles well, but using the `iron` crate fails:

```
error: internal compiler error: static call to invalid vtable: VtableObject(VtableObject(object_ty=error::Error + 'static))

   7:     0x7f6f80c991f0 - trans::meth::trans_static_method_callee::h7b45cabbb9a118b8sbz
   8:     0x7f6f80c960d0 - trans::callee::trans::h36851c362b65c790yag
   9:     0x7f6f80ca02f0 - trans::callee::trans_call_inner::h8685402877239769531
  10:     0x7f6f80ca5d10 - trans::expr::trans_rvalue_dps_unadjusted::hadddfbae4195cb641Si
  11:     0x7f6f80c5c5f0 - trans::expr::trans_into::ha60adcbe08fa3bdbKyh
```

**Comment by nikomatsakis** [comment]
Fix in #21089, sorry for the delay.

**Comment by shepmaster** [comment]
@nikomatsakis thanks for the work! I pulled your PR and built with it, and my code compiles again. Hooray!
